File: ACM/NAQualifier2017/B.py
import queue
n,m,f,s,t = [int(x) for x in input().split()]
graph = {}
for a0 in range(m):
    i,j,c = [int(x) for x in input().split()]
    if i in graph:
        graph[i].append((j,c))
    else:
        graph[i] = [(j,c)]
    if j in graph:
        graph[j].append((i,c))
    else:
        graph[j] = [(i,c)]

graphf = {}
for a1 in range(f):
    u, v = [int(x) for x in input().split()]
    graphf[u] = v
    
if f > 0:
    minn = []
    for a2 in graphf:
        u = a2
        v = graphf[a2]
        newgraph = dict(graph)
        for a3 in newgraph[u]:
            if a3[0] == v:
                newgraph[u].remove(a3)
                break
        for a3 in newgraph[v]:
            if a3[0] == u:
                newgraph[v].remove(a3)
                break
        newgraph[u].append((v,0))
        newgraph[v].append((u,0))
        #print(newgraph) 
        t1 = [50001]*n
        pq = queue.PriorityQueue()
        
        #unvisited = {node: None for node in range(n)} 
        pq.put((0,s,s))
        #unvisited[s] = 0
        #print(unvisited)
        visited = set()
        while True:
            if t in visited:
                break
            #candidates = [node for node in unvisited.items() if node[1] or node[1]==0]
            #print(candidates) 
            
            #newnode, curcost = sorted(candidates, key = lambda x: x[1])[0]
            newnode = pq.get()
            curcost = newnode[0]
            newnode = newnode[1]
            for a4 in newgraph[newnode]:
                dest = a4[0]
                cost = a4[1]
                newcost = curcost + cost
                if dest in visited:
                    continue
                if newcost < t1[dest]:
                    t1[dest] = newcost
                    #unvisited[dest] = newcost
            #del unvisited[newnode]
            visited.add(newnode)
            #if not unvisited:
            #    break
            
        minn.append(t1[t])
    print(min(minn))

else:
    t1 = [50001]*n
    pq = queue.PriorityQueue()
    
    unvisited = {node: None for node in range(n)} 
    #pq.put((0,s,s))
    unvisited[s] = 0
    #print(unvisited)
    visited = set()
    while True:
        if t in visited:
            break
        candidates = [node for node in unvisited.items() if node[1] or node[1]==0]
        #print(candidates) 
        newnode, curcost = sorted(candidates, key = lambda x: x[1])[0]
        #curcost = newnode[0]
        for a4 in graph[newnode]:
            dest = a4[0]
            cost = a4[1]
            newcost = curcost + cost
            if dest not in unvisited:
                continue
            if newcost < t1[dest]:
                t1[dest] = newcost
                unvisited[dest] = newcost
        del unvisited[newnode]
        visited.add(newnode)
        if not unvisited:
            break
        
    print(t1[t])

File: ACM/NAQualifier2017/B.cpp
#include <iostream>
#include <unordered_map>
#include <utility>
#include <queue>
using namespace std;
int main()
{
	int n,m,f,s,t;
	cin>>n>>m>>f>>s>>t;
	cout<<n<<m<<f<<s<<t;
	unordered_map<int,unordered_map<int,int> > graph;
	
	for (int a0 =0;a0<m;++a0)
	{
		int i,j,c;
		cin>>i>>j>>c;
		//if (graph.count(i)){
			graph[i][j] = c;
		graph[j][i] = c;
			//}
		//else
		//{
		//	graph[i] = unordered_map<int,int> sgraph;
		//}		
	}
	unordered_map<int, int> graphf;
	for (int a0 = 0;a0<f;++a0)
	{
		int u,v;
		cin>>u>>v;
		graphf[u] = v;		
	}
	if (f>0)
		int minn = 50001;
		
		for (auto const& a2:graphf)
		{
			int u = a2.first;
			int v = a2.second;
			unordered_map<int,unordered_map<int,int> > newgraph;
			
			newgraph = graph;
			//cout<<newgraph[1][3]<<endl;
			newgraph[u][v] = 0;
			newgraph[v][u] = 0;
			//cout<<newgraph[1][3]<<endl;
		}
		vector<int> t1(n,minn);
		pq = 
		 	

}
